### ISSUE TITLE
Create iOS IPA from command line

### ISSUE THREAD

**Issue by passsy** [flutter/flutter#13065]
Flutter doesn't allow me to build an .ipa which can be published to hockeyapp/fabric or the app store.

```bash
# creates an .app not an .ipa 
flutter build ios --release
```

Manual bundling is required:
```bash
xcodebuild -exportArchive -archivePath build/ios/iphoneos/Runner.app -exportPath build/ios/iphoneos/ -exportOptionsPlist ci/enterprise-options.plist
```

Flutter includes tools such as measuring the ipa size https://github.com/flutter/flutter/issues/4609, but I can't find a way to create an .ipa.

A followup issue for [Generate a production-ready IPA from a single command](https://github.com/flutter/flutter/issues/1417) is still missing

**Comment by eseidelGoogle** [comment]
I don't know the details, but if it's really a simple as just one command we could certainly wrap that in `flutter build ipa` or similar. :)  I had previously been under the impression that the XCode GUI had to be involved with .ipa generation (e.g. it wasn't possible from the command line).  @cbracken or @xster would know more.

**Comment by sethladd** [comment]
cc @mit-mit @devoncarew for CLI

**Comment by xster** [comment]
cc @cbraken

**Comment by mit-mit** [comment]
@cbracken can you elaborate?

**Comment by cbracken** [comment]
Thanks for filing this issue!

Our currently supported iOS releasing process is described in [Releasing for iOS](https://flutter.io/ios-release/) and as noted, makes use of the Xcode UI. We don’t yet support a pure command-line mechanism for releasing to the App Store, and it’s not something we currently have prioritised but I’d certainly be more than happy to review pull requests.

**Comment by oemera** [comment]
@cbracken thanks for your response. The error also happens when archiving within the Xcode UI. Please look into my issue: #13070. Based on the reactions I guess there are a couple of others who have this issue and can't release their app with the new version of Xcode.

If you have any questions please let me know.

**Comment by cbracken** [comment]
Thanks for pointing to #13070, I'll take a look at that today.

**Comment by alibitek** [comment]
[xcodebuild: Deploy iOS app from Command Line](https://medium.com/xcblog/xcodebuild-deploy-ios-app-from-command-line-c6defff0d8b8)
A really good summary of xcodebuild command line capabilities.
It even includes an example of how you can submit the .ipa to iTunes Connect from the command line.

**Comment by peace2knowledge** [comment]
change to project ios directory,excute flowing command:

```sh
xcodebuild -workspace Runner.xcworkspace -scheme Runner -sdk iphoneos -configuration Release archive -archivePath $PWD/build/Runner.xcarchive

xcodebuild -exportArchive -archivePath $PWD/build/Runner.xcarchive -exportOptionsPlist exportOptions.plist -exportPath $PWD/build/Runner.ipa
```
exportOptions.plist is on ios directory.
file content like this:
``` xml 
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>compileBitcode</key>
    <true/>
    <key>method</key>
    <string>enterprise</string>
    <key>provisioningProfiles</key>
    <dict>
        <key>com.company.appname</key>
        <string>provisioningProfiles name </string>
    </dict>
    <key>signingCertificate</key>
    <string>iOS Distribution</string>
    <key>signingStyle</key>
    <string>manual</string>
    <key>stripSwiftSymbols</key>
    <true/>
    <key>teamID</key>
    <string>xxxxxx</string>
    <key>thinning</key>
    <string>&lt;none&gt;</string>
</dict>
</plist>
```
@oemera

**Comment by xster** [comment]
This is theoretically possible. I got it to work using Fastlane (which really just invokes xcodebuild archive). 

We can document the flow more or possibly integrate Fastlane into flutter tools, though there are no immediate plans yet. 

In the meantime, you can `flutter build ios --release --no-codesign` first and then use `build_ios_app(export_method: "app-store", ...)` in Fastlane with whichever signing method you prefer.

**Comment by csga5000** [comment]
+1 on this.  I think building iPa's would be great.  During beta phase of apps I sometimes have to release new versions every couple days, and it takes forever when I have to test the app by running it on a device, then do flutter build --release, and then open xcode, and do product archive, and then in organizer upload to app store, and finally wait for apple to process my app.

I think this could be streamlined to just: `flutter build ipa --release` and then open application loader and deliver my app, (and then still wait for apple to process the app).  It's still a bit slow, but at least I don't have to wait forever for flutter to build my app so I can wait another forever for xcode to archive my app.  Admittedly it may be so slow partially cause my mac (not my main machine) is a mac mini and runs like a slug.

If you're not aware of application loader, you can open it through xcode *XCode > Open Developer Tool > Application Loader*, but you can also pin it to your task bar to open it without opening xcode (or install it separately, though that generally is very out of date).

**Comment by fuzzybinary** [comment]
I would also appreciate this since we want to build ipas for distrubution through CI.

One (fairly crazy) issue I'm having with the current build system is that changing between `flutter build ios --debug` and `flutter build ios --release` actually modifies the Podfile.lock, and requires a fresh pod install, which is overriding a change I need to make in one of our Pods to avoid https://github.com/grpc/grpc/issues/7406

Also because of this, we have to run `flutter build ios --release` followed by `xcodebuild archive` followed by `xcodebuild exportArchive` to build an ipa. This causes the app to be built twice (kind of a waste of time).

It seams like a lot of the build system could use refactoring. In general:

- You should be able to generate ipas for a given flavor from the command line
- flavors should be supported more directly from flutter cli (see https://github.com/flutter/flutter/issues/14648)
- switching flavors shouldn't modify the Podfile.lock. This may require modifying Debug/Release.xcconfig to not rely as much on the Generated.xcconfig (or at least removing the location of the Flutter SDK from Generated and putting it in Debug/Release.xcconfig

I'm actually more than willing to help with this with some direction from the flutter team on what they think the best approach would be.

**Comment by eseidelGoogle** [comment]
CC @mravn-google

**Comment by neelu004** [comment]
Hi, I am using Mac and in cmd I am running flutter build ios this command which will generates my Runner.app in build folder. I want to know how to run this Runner.app in real iOS device.
Please suggest some solution.
Thanks!

**Comment by MisterJimson** [comment]
@neelu004 check the docs, this will get you started. https://flutter.dev/docs/get-started/test-drive?tab=terminal

**Comment by Helios94** [comment]
@neelu004 if you successfully generated the .ipa file, you can install it on your iOS device using Cydia

**Comment by castrojr913** [comment]
In addtion to @MisterJimson response, it's a need to create `exportOptions.plist` file on iOS folder. The content of this file depends on the distribution type (appstore or ad-hoc). For instance, if you need to deploy to Ad-hoc, this file could be:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
    <dict>
        <key>provisioningProfiles</key>
        <dict>
            <key>your_bundle_id</key>
            <string>your_provisioning_profile_name</string>
        </dict>
        <key>method</key>
        <string>ad-hoc</string>
        <key>teamID</key>
        <stringyour_team_id</string>
        <key>uploadBitcode</key>
        <false/>
        <key>uploadSymbols</key>
        <true/>
    </dict>
</plist>
```

**Comment by Levi-Lesches** [comment]
Any official progress on this? I am interested in using [Firebase app distribution](https://firebase.google.com/products/app-distribution), and being able to just run `flutter build ios` (as one would just run `flutter build apk`) would be great.

**Comment by chrisdrobison** [comment]
I'm not sure this applies as I'm very new, but `flutter build ios` creates a .app folder. From what I read, that can be converted to an IPA: https://bitbar.com/blog/how-to-create-ipa-and-xctest-package-for-ios-testing/

**Comment by Levi-Lesches** [comment]
I too don't know much about iOS but I'm pretty sure that has to be re-signed (or basically whatever Xcode is doing when you press "Archive").

**Comment by MisterJimson** [comment]
@chrisdrobison Yes that's how we have to make IPAs today. First that command, then some xcode CLI commands (or the xcode UI) to convert to an IPA.

This issue is about introducing the ability to create an IPA directly, without subsequent commands or steps.

**Comment by chrisdrobison** [comment]
@MisterJimson I guess I was referring specifically to the part of the article where he is just zipping a .app folder and calling it .ipa. Flutter already outputs a .app folder. So either I'm missing something, but it seems like I don't have to call xcodebuild at all.

**Comment by therealjohnsummer** [comment]
@chrisdrobison To just get the IPA from the build so that in can be dropped onto CI servers like AppCenter  - this process will work : 

![image](https://user-images.githubusercontent.com/54981654/68071070-304b9380-fd6e-11e9-80ca-779ee35f0b48.png)

**Comment by algodave** [comment]
@therealjohnsummer I don't know the complexity of the app you're working at, but in my case what you described is just not viable, because the `Runner.app` outputted by `flutter build ios --release` is not acceptable in terms of size: about 570 MB. And even if I zip it, I only save a few KBytes.

**Comment by therealjohnsummer** [comment]
@algodave My .app file size for debug is 450MB - when I follow the steps outlined above , and **compress** the payload folder and rename to iPA - the file size is approx 56.8mb which is ok for me to drop into CI servers etc  - did you follow the steps correctly above ?

**Comment by algodave** [comment]
@therealjohnsummer I am following steps shared by @MisterJimson using `xcodebuild` command and I personally think that it's the best answer provided in this issue so far. 

I really hope that `flutter build ios` will soon cover those steps, especially for coherence with the Android side, where everything works "automagically" behind the scenes.

**Comment by therealjohnsummer** [comment]
@algodave Ok - so steps are not the same then. But I completely agree with you. To not have a command 'flutter build ipa --release | --debug' built into the Flutter CLI tools is a huge omission.  I would **LOVE** to see flutter fully support IPA creation within flutter toolset.

**Comment by Levi-Lesches** [comment]
@therealjohnsummer The only problem I have with this approach (asides from Flutter not doing all the work for me) is that when I set up a Powershell script (using the built-in zipper) to do just this, I got errors I couldn't install it. I used 7-Zip manually and that works, though.

**Comment by szotp** [comment]
It would be nice to have option execute "flutter build ios" without running xcodebuild, this way we can switch to fastlane gym without wasting time building twice. 

It would also be pretty easy change.

**Comment by shouichi** [comment]
We can create IPA by just zipping `Runner.app`.

```
$ flutter build ios --release
$ mkdir -p build/ios/iphoneos/Payload
$ mv build/ios/iphoneos/Runner.app build/ios/iphoneos/Payload
$ cd build/ios/iphoneos
$ zip -r app.ipa Payload
```

**Comment by Levi-Lesches** [comment]
The problem with this is that it doesn't handle signing. As per @MisterJimson: 
> That may work for running XCTest tests, but for AppStore, Ad Hoc, and Enterprise releases you need to sign the app when converting to an IPA. There are examples of how to to that in this thread. It's done automatically for Android apps, but not yet for iOS apps.

**Comment by szotp** [comment]
@shouichi We had hack that used this but it was very bad idea. The ipa should be prepared from xcodebuild archive (fastlane gym does it internally), otherwise you are asking for trouble.

**Comment by chrisdrobison** [comment]
Yeah, I tried getting this to work as well, but it did not end well. I finally got something figured out in Azure DevOps to build and signn.

**Comment by shouichi** [comment]
@szotp Thanks for the advice. Yeah, I agree that it's not a very good idea. Feels like depending on the `xcodebuild` internal which can change anytime.

What's preventing `flutter build ios` command from building IPA file anyway?

**Comment by MisterJimson** [comment]
Nothing but the work required to do it, and some documentation around creating a `exportOptions.plist`

**Comment by lukaselmer** [comment]
In addtion to @MisterJimson and @castrojr913, you can also use the flag `-allowProvisioningUpdates` to handle signing automatically.

```sh
flutter build apk
flutter build ios
rm -rf build/export
mkdir -p build/export
cp build/app/outputs/apk/release/app-release.apk build/export
cd ios
xcodebuild -workspace Runner.xcworkspace -scheme Runner -sdk iphoneos -configuration Release archive -archivePath $PWD/build/Runner.xcarchive
xcodebuild -exportArchive -archivePath $PWD/build/Runner.xcarchive -exportOptionsPlist ExportOptions.plist -exportPath $PWD/build/Runner.ipa -allowProvisioningUpdates
cd ..
cp ios/build/Runner.ipa/Runner.ipa build/export
open build/export
```

with ios/ExportOptions.plist

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
    <dict>
        <key>method</key>
        <string>ad-hoc</string>
        <key>uploadBitcode</key>
        <false/>
        <key>uploadSymbols</key>
        <true/>
    </dict>
</plist>
```

**Comment by sallymariehollywood** [comment]
@lukaselmer  This looks very promising what you have put here (Specifically for IPA) and will try it out ... But before I do :- 

1) Does this assume Adhoc provisioning profiles/Certificates have been created in developer.apple.com ? Or will this do this for me AUTOMATICALLY ?  I'm after a quick way to release Adhoc builds to testFlight for beta test users ... 

2) Your creating 'release' ipa builds, will this work for DEBUG builds ?

**Comment by lukaselmer** [comment]
@sallymariehollywood 

> 1. Does this assume Adhoc provisioning profiles/Certificates have been created in developer.apple.com ? Or will this do this for me AUTOMATICALLY ?  I'm after a quick way to release Adhoc builds to testFlight for beta test users ...

You have to enable automatic singing in Xcode first:

![image](https://user-images.githubusercontent.com/326935/72216918-bc74f680-3527-11ea-95a8-668b6ca32bc1.png)

It will guide you through provisioning profile / certificate creation.

> 2. Your creating 'release' ipa builds, will this work for DEBUG builds ?

Hmm, I don't know. Maybe you could try `xcodebuild -workspace Runner.xcworkspace -scheme Runner -sdk iphoneos -configuration Debug archive -archivePath $PWD/build/Runner.xcarchive` to create a `Debug` build. 🤔

**Comment by Venusdjinni** [comment]
> In addtion to @MisterJimson and @castrojr913, you can also use the flag `-allowProvisioningUpdates` to handle signing automatically.
> 
> ```shell
> flutter build apk
> flutter build ios
> rm -rf build/export
> mkdir -p build/export
> cp build/app/outputs/apk/release/app-release.apk build/export
> cd ios
> xcodebuild -workspace Runner.xcworkspace -scheme Runner -sdk iphoneos -configuration Release archive -archivePath $PWD/build/Runner.xcarchive
> xcodebuild -exportArchive -archivePath $PWD/build/Runner.xcarchive -exportOptionsPlist ExportOptions.plist -exportPath $PWD/build/Runner.ipa -allowProvisioningUpdates
> cd ..
> cp ios/build/Runner.ipa/Runner.ipa build/export
> open build/export
> ```
> 
> with ios/ExportOptions.plist
> 
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
> <plist version="1.0">
>     <dict>
>         <key>method</key>
>         <string>ad-hoc</string>
>         <key>uploadBitcode</key>
>         <false/>
>         <key>uploadSymbols</key>
>         <true/>
>     </dict>
> </plist>
> ```
> 
> https://gist.github.com/lukaselmer/2a2ac2e9be77a0fd66168610242c5ed0

I'm ending up with this error when running the exportArchive command:

2020-01-13 18:04:55.872 xcodebuild[45634:317859] [MT] IDEDistribution: -[IDEDistributionLogging _createLoggingBundleAtPath:]: Created bundle at path '/var/folders/09/rgbfggn51xg0505lzhkpmv_00000gn/T/Runner_2020-01-13_18-04-55.871.xcdistributionlogs'.
2020-01-13 18:05:03.216 xcodebuild[45634:317917]  DVTAssertions: ASSERTION FAILURE in /Library/Caches/com.apple.xbs/Sources/IDEFrameworks/IDEFrameworks-15712/IDEFoundation/Provisioning/Mechanic/IDEProvisioningRepairStepGenerator.m:252
Details:  Didn't find a single primary certificate type from {(
)}
Object:   <IDEProvisioningRepairStepGenerator_Automatic: 0x7fc4a6c0eaa0>
Method:   -_evaluateCertificatesWithSession:context:repairable:steps:userAction:error:
Thread:   <NSThread: 0x7fc4ab937470>{number = 4, name = (null)}
Hints: 

Backtrace:
  0   -[DVTAssertionHandler handleFailureInMethod:object:fileName:lineNumber:assertionSignature:messageFormat:arguments:] (in DVTFoundation)
  1   _DVTAssertionHandler (in DVTFoundation)
  2   _DVTAssertionFailureHandler (in DVTFoundation)
  3   -[IDEProvisioningRepairStepGenerator_Automatic _evaluateCertificatesWithSession:context:repairable:steps:userAction:error:] (in IDEFoundation)
  4   -[IDEProvisioningRepairStepGenerator_Automatic stepsWithSession:context:userAction:error:] (in IDEFoundation)
  5   -[IDEProvisioningRepair execute] (in IDEFoundation)
  6   __53-[IDEProvisioningMechanic _executeNextRepairIfNeeded]_block_invoke_2 (in IDEFoundation)
  7   __DVT_CALLING_CLIENT_BLOCK__ (in DVTFoundation)
  8   __DVTDispatchAsync_block_invoke (in DVTFoundation)
  9   _dispatch_call_block_and_release (in libdispatch.dylib)
 10   _dispatch_client_callout (in libdispatch.dylib)
 11   _dispatch_continuation_pop (in libdispatch.dylib)
 12   _dispatch_async_redirect_invoke (in libdispatch.dylib)
 13   _dispatch_root_queue_drain (in libdispatch.dylib)
 14   _dispatch_worker_thread2 (in libdispatch.dylib)
 15   _pthread_wqthread (in libsystem_pthread.dylib)
 16   start_wqthread (in libsystem_pthread.dylib)
zsh: abort      xcodebuild -exportArchive -archivePath $PWD/build/Runner.xcarchive

**Comment by lukaselmer** [comment]
@Venusdjinni did you create a certificate at

 * [https://developer.apple.com/account/resources/certificates/list](https://developer.apple.com/account/resources/certificates/list) 
* and a provisioning profile at [https://developer.apple.com/account/resources/profiles/list](https://developer.apple.com/account/resources/profiles/list)

?

**Comment by Venusdjinni** [comment]
I used an Apple generated one. It is showed in the Xcode GUI in Signing and Capabilities

**Comment by ArkFreestyle** [comment]
> We can create IPA by just zipping `Runner.app`.
> 
> ```
> $ flutter build ios --release
> $ mkdir -p build/ios/iphoneos/Payload
> $ mv build/ios/iphoneos/Runner.app build/ios/iphoneos/Payload
> $ cd build/ios/iphoneos
> $ zip -r app.ipa Payload
> ```

Can someone let me know WHY this is a bad idea?

**Comment by MisterJimson** [comment]
@ArkFreestyle that doesn't properly archive the app into an IPA. Upload it to Apple for TestFlight/App Store release and you will run into errors.

**Comment by avi-mastov** [comment]
> > We can create IPA by just zipping `Runner.app`.
> > ```
> > $ flutter build ios --release
> > $ mkdir -p build/ios/iphoneos/Payload
> > $ mv build/ios/iphoneos/Runner.app build/ios/iphoneos/Payload
> > $ cd build/ios/iphoneos
> > $ zip -r app.ipa Payload
> > ```
> 
> Can someone let me know WHY this is a bad idea?

@ArkFreestyle 
If you unzip a real IPA created by xcode, you will see it contains many more files than just Runner.app.

 The need to flutter build and then archive is really a pain that makes the build run twice the time, because xcode and flutter use different derived data folders and the archive process can't find the intermediate files that were created by flutter build. All my attempts to set the derived data paths to be the same have failed.

**Comment by ArkFreestyle** [comment]
@MisterJimson @avi-mastov ah I see, so we only run into errors if we try releasing that to the app store, but otherwise the IPA will still work on ios devices (for test purposes etc) right?

**Comment by MisterJimson** [comment]
@ArkFreestyle I'm unsure if that IPA can be loaded onto a device. I only ever try with properly signed apps.

**Comment by ArkFreestyle** [comment]
> @ArkFreestyle I'm unsure if that IPA can be loaded onto a device. I only ever try with properly signed apps.

Hi again, I tried shouichi's method and zipped Runner.app into an IPA, it does indeed work on an iphone. Just thought I'd post an update here for all to see.

**Comment by nagarajsh-dev** [comment]
tried creating the spa file from a Payload folder zip, worked fine on my college's device who had provisioning profile already on his Mac, but failed to download on another colleague without provisioning profile, saying "A valid provisioning profile for this executable was not found"... I did have sign automatically set on my Xcode. I seriously think we need flutter solution for building ipa file. when I try to Archive on Xcode after running "flutter build ios --release", I get error on Ruby file. looks like Xcode uses different version than what flutter uses (cocoa pods). any suggestions are highly appreciated.

**Comment by Alekcei** [comment]
please add 
```flutter build ipa --release --ad-hoc```
```flutter build ipa --release --app-store```

**Comment by basharh** [comment]
This would definitely make life much easier when implementing CI/CD pipelines.

**Comment by timshoaf** [comment]
Would love to have this tooling. Digging through various documentation to find the various magic incantations / wards necessary for handling this all on command line is laborious. The CI / CD pipeline use case is very real for small dev shops. Absolutely +1.

**Comment by jjones-figure** [comment]
+1 we need this. Currently we have to run `flutter build ios` and then manually execute an archive, which essentially doubles our build time on CI/CD (60+ minutes!).

**Comment by xster** [comment]
@jonahwilliams has a draft at https://github.com/flutter/flutter/pull/64683

**Comment by jonahwilliams** [comment]
So in order to land that PR, a few things would need to be done:

I'd need take a pass through the existing flutter build ios and perform a bit of cleanup, so that subsequent PRs have any easier time sharing code and writing new tests. Then we could add (perhaps hidden behind a flag) `flutter build ipa` with a minimal configuration like above. To make it usable we'd need a handful of interested users to give it a test run before we called it good.

Separately, we could introduce a `--config-only` flag to flutter build ios, to at least skip the double building for users that are already using fastlane/xcode to build

**Comment by xster** [comment]
ya, that's the biggest challenge for this. Conceptually code signing works very differently than any of our existing work in the code_signing.dart tool since this needs a distribution certificate rather than a development certificate. We can either deeply integrate with App Store Connect a bit like https://docs.fastlane.tools/actions/match/ does (very difficult) or create something between `flutter pub get` and `flutter build ios [add all the flags to get your xcconfig ready]` as suggested. 

We can also create a hybrid workflow where we have some manual instructions on the website to have users get their distribution certificate ready and provisioning profile created, then let `flutter build ipa` be pointed to the provisioning profile.

**Comment by szotp** [comment]
xcodebuild archive has huge surface area so it's basically impossible to do something that works for everyone. For me, just adding `--config-only` would do the job, but for others it could be enough to provide a predefined script that works for 80% of cases (automatic signing probably) but can also be edited manually (like the Podfile).

**Comment by MisterJimson** [comment]
I suggest `flutter build ipa` that requires a config file like `exportoptions.plist`.

Similar in some ways to how users need to manually create key stores and set passwords before building an Android release build that is ready for Google Play.

So the only manual part would be to create the plist file. I feel like deep integration with App Store Connect is out of scope for this issue, or at least the first pass of it.

**Comment by jonahwilliams** [comment]
@MisterJimson thats a good suggestion for an initial pass - we could support accepting the plist via command line parameter. 

I don't know if it would also be useful to support generating the config from command line options, since that would probably take a bit more research on my end. Bear with me, because I've literally never released software on the App Store, but is it common to generate the export options on the fly with other software, or is it something that is treated as a secret like the Android keystore

**Comment by xster** [comment]
Another option we have is to still break out the current `flutter build ios` into a `flutter build get my xcconfig ready for xcode` and a `flutter build actually xcodebuild` but for better ergonomics, integrate `flutter build get my xcconfig ready for xcode` into a Flutter fastlane build action https://github.com/flutter/flutter/issues/17402 or document how to follow up with https://docs.fastlane.tools/actions/gym/ directly

**Comment by jonahwilliams** [comment]
@xster that is what the suggestion for --config-only basically is. Though at the point where you are archiving through another tool there isn't anything else for the flutter CLI to do.

**Comment by xster** [comment]
ya, we're saying the same thing.

**Comment by MisterJimson** [comment]
@jonahwilliams I've definitely seem both ways. I generally see 2 types of approaches that companies use, depending on the use case and preference. (And to be clear this is just for Ad hoc / App Store release, not day to day debugging)

Fastlane, which I think Flutter should integrate with or provide some type of easy path for usage at some point. I feel it would be difficult to replicate and maintain a similar feature set in Flutter.

Static `exportoptions.plist` that scripts pass to xcodebuild.

**Comment by jonahwilliams** [comment]
I think its unlikely we'll consider deeper Fastlane integration in the short-term. Integration with any sort of 3rd party service tends to have a high maintenance cost, both in terms of the actual integration, updates and the inevitable support requests ( mostly involving the service itself) which will start getting routed through our issue tracker.

We should only do so where absolutely necessary - i.e. we integrate with Xcode even though it is a pain because that is existential for iOS development.

**Comment by avi-mastov** [comment]
I'm totally in favor of breaking `flutter build ios` into `--config-only` and the actual build so developers could then start xcode archive immediately and build only once. 

That would require very little change to current developers' build process / CI/CD while speeding it up significantly, and on the other hand won't require any integration on Flutter's part.

**Comment by jonahwilliams** [comment]
Would love if someone wanted to checkout + sanity check https://github.com/flutter/flutter/pull/64848 :)

**Comment by Pe-te** [comment]
Just a crazy idea after reading everything: Would it be possible to only run "fastlane gym" and add a build step to the project that does the "flutter --config-only" stuff? Or will this create all kinds of other problems?

**Comment by xster** [comment]
Not crazy. That's the intent here. Once you flutter --config-only, you can then build the iOS app via standard iOS mechanisms such as `xcodebuild` if you do so in your CI. Or just run or archive from Xcode UI, or use fastlane.

**Comment by Pe-te** [comment]
That would be perfect, thanks! I first had understood that we still need to call it before xcodebuild.

**Comment by szotp** [comment]
Seems like @xster misunderstood you. It's not possible, because --config-only changes some variables that affect the xcode project, and it also takes parameters that need to be be specified somehow, like entry file or dart defines.

**Comment by xster** [comment]
@Pe-te I think you're asking if you can just build with `flutter build ios --config-only` then `fastlane gym` right? Yes. If you run fastlane gym in verbose mode, I think it just wraps xcodebuild calls. 

@szotp I might have misunderstood :) what do you mean by "it also takes parameters that need to be be specified somehow, like entry file or dart defines"? If you meant things like `flutter run --target` etc, they're actually Xcode project variables behind the scene as well.

**Comment by szotp** [comment]
> Would it be possible to only run "fastlane gym" and add a build step to the project that does the "flutter --config-only" stuff?

@xster Pe-te was asking if it was possible to automate `flutter build ios --config-only` as a build step inside xcodeproj, so that the only thing needed is `fastlane gym`. But that's not possible because it would mean changing xcodeproj while you are building it.

**Comment by xster** [comment]
Ah I misunderstood. You mean rather than in your CI script or in your Fastfile, whether you can recursively invoke `flutter build ios --config-only` inside Xcode itself with a build phase? 

(Not tested) I think that works if you invoked it early enough before the build Flutter phase or the compile sources phase but doesn't that defeat the point? In other words, the whole point is to tell Xcode what configuration to build and archive etc. If Xcode already knows, then it could just hardcode those values from `Generated.xcconfig` (which is more or less all `flutter build ios --config-only` does) in your .xcodeproj directly.

**Comment by Pe-te** [comment]
Thanks for all the clarifications, just as an addition why I asked:

Right now I have in my buildscript:

fastlane init_keychain
flutter build  (soon --config-only)
fastlane build_my_app
fastlane close_keychain

So it would be nice to reduce it to "fastlane build_my_app" and do all the other stuff inside the Fastfile to have the fastlane error handling for everything. I know, I could call flutter through a fastlane sh() call, but that feels weird. 😉 

So the idea was to pack the --config-only also into the build phases, but I am not sure what it does exactly and how it affects the build process if you change stuff while it runs.

**Comment by jmagman** [comment]
Can we track the `--config-only` flag in a different feature request issue?  I only just saw this, and have a few concerns, but I don't want to further muddy this feature request.

As for `flutter build ipa` and why it isn't just done
There are so many options when you archive that would have to be exposed in that command just to pipe 1:1 into `xcodebuild`/`exportOptionsPlist`.  You pretty much need an entire UI wizard flow in the CLA to make it manageable, which Xcode already has (and keeps updated with new App Store requirements).  So either we build an identical wizard flow, or we expose all the below flags, and if you can successfully use them, then congrats! you are an Xcode guru and are probably already using `xcodebuild` directly.

```
xcodebuild -exportArchive -archivePath <xcarchivepath> [-exportPath <destinationpath>] -exportOptionsPlist <plistpath>
```
```
Available keys for -exportOptionsPlist:

	compileBitcode : Bool

		For non-App Store exports, should Xcode re-compile the app from bitcode? Defaults to YES.

	destination : String

		Determines whether the app is exported locally or uploaded to Apple. Options are export or upload. The available options vary based on the selected distribution method. Defaults to export.

	distributionBundleIdentifier : String

		Reformat archive to focus on eligible target bundle identifier.

	embedOnDemandResourcesAssetPacksInBundle : Bool

		For non-App Store exports, if the app uses On Demand Resources and this is YES, asset packs are embedded in the app bundle so that the app can be tested without a server to host asset packs. Defaults to YES unless onDemandResourcesAssetPacksBaseURL is specified.

	generateAppStoreInformation : Bool

		For App Store exports, should Xcode generate App Store Information for uploading with iTMSTransporter? Defaults to NO.

	iCloudContainerEnvironment : String

		If the app is using CloudKit, this configures the "com.apple.developer.icloud-container-environment" entitlement. Available options vary depending on the type of provisioning profile used, but may include: Development and Production.

	installerSigningCertificate : String

		For manual signing only. Provide a certificate name, SHA-1 hash, or automatic selector to use for signing. Automatic selectors allow Xcode to pick the newest installed certificate of a particular type. The available automatic selectors are "Mac Installer Distribution" and "Developer ID Installer". Defaults to an automatic certificate selector matching the current distribution method.

	manifest : Dictionary

		For non-App Store exports, users can download your app over the web by opening your distribution manifest file in a web browser. To generate a distribution manifest, the value of this key should be a dictionary with three sub-keys: appURL, displayImageURL, fullSizeImageURL. The additional sub-key assetPackManifestURL is required when using on-demand resources.

	method : String

		Describes how Xcode should export the archive. Available options: app-store, validation, ad-hoc, package, enterprise, development, developer-id, and mac-application. The list of options varies based on the type of archive. Defaults to development.

	onDemandResourcesAssetPacksBaseURL : String

		For non-App Store exports, if the app uses On Demand Resources and embedOnDemandResourcesAssetPacksInBundle isn't YES, this should be a base URL specifying where asset packs are going to be hosted. This configures the app to download asset packs from the specified URL.

	provisioningProfiles : Dictionary

		For manual signing only. Specify the provisioning profile to use for each executable in your app. Keys in this dictionary are the bundle identifiers of executables; values are the provisioning profile name or UUID to use.

	signingCertificate : String

		For manual signing only. Provide a certificate name, SHA-1 hash, or automatic selector to use for signing. Automatic selectors allow Xcode to pick the newest installed certificate of a particular type. The available automatic selectors are "Mac App Distribution", "iOS Developer", "iOS Distribution", "Developer ID Application", "Apple Distribution", "Mac Developer", and "Apple Development". Defaults to an automatic certificate selector matching the current distribution method.

	signingStyle : String

		The signing style to use when re-signing the app for distribution. Options are manual or automatic. Apps that were automatically signed when archived can be signed manually or automatically during distribution, and default to automatic. Apps that were manually signed when archived must be manually signed during distribtion, so the value of signingStyle is ignored.

	stripSwiftSymbols : Bool

		Should symbols be stripped from Swift libraries in your IPA? Defaults to YES.

	teamID : String

		The Developer Portal team to use for this export. Defaults to the team used to build the archive.

	thinning : String

		For non-App Store exports, should Xcode thin the package for one or more device variants? Available options: <none> (Xcode produces a non-thinned universal app), <thin-for-all-variants> (Xcode produces a universal app and all available thinned variants), or a model identifier for a specific device (e.g. "iPhone7,1"). Defaults to <none>.

	uploadBitcode : Bool

		For App Store exports, should the package include bitcode? Defaults to YES.

	uploadSymbols : Bool

		For App Store exports, should the package include symbols? Defaults to YES.
```
And that will just build the archive, it doesn't run any validators.

**Comment by jmagman** [comment]
> Can we track the `--config-only` flag in a different feature request issue? I only just saw this, and have a few concerns, but I don't want to further muddy this feature request.

Never mind, it was already merged :-(

**Comment by jmagman** [comment]
The value add from `flutter` tooling would be the creation of the `xcarchive`--that's what is going to use all the build settings, compiling dart/embedding frameworks/thinning logic.  So a `flutter build xcarchive` command.

The last step of making the `ipa` is entirely specific to the developer and what they are using the `ipa` for, and Flutter can't and shouldn't assume any of the `exportOptionsPlist` values.  So the developer is already signed up with reading some Apple docs to get the right values, and at that point the value of running
```
xcodebuild -exportArchive -archivePath <xcarchivepath> [-exportPath <destinationpath>] -exportOptionsPlist <plistpath>
```
vs
```
flutter build ipa -exportPath <destinationpath> -exportOptionsPlist <plistpath>
```
is not that valuable (though we could add that wrapper, it's just not the hard part).

**Comment by jmagman** [comment]
Created https://github.com/flutter/flutter/issues/66619 to discuss there.

**Comment by MisterJimson** [comment]
@jmagman I see your point but instead of mapping all of the exportOptionsPlist options we could just take a path to that file. Similar to how your keystore needs to exist on disk and how you have to manually setup your build.gradle file before you want to release for Android. There can be a happy medium for `flutter build ipa`.

**Comment by jmagman** [comment]
[`flutter build xcarchive`](#67598) is now available on the [master channel](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#how-to-change-channels).

**Comment by jmagman** [comment]
Command has been renamed `flutter build ipa` (alias `flutter build xcarchive` still works) and takes an optional `--export-options-plist` flag to make an IPA from the archive bundle.  See `flutter build ipa -h` for details.
Website documentation forthcoming at https://github.com/flutter/website/issues/4852.

**Comment by jmagman** [comment]
If you're still interested in building for distribution from the command line...

`flutter build ipa` has two stages for iOS distribution:
1. Archive the app into an `xcarchive` bundle.
2. Optionally, export the `xcarchive` into an `ipa` bundle for distribution to the App Store, or to a device, or enterprise distribution...

Currently to get to number 2 and create the `ipa` file, users need to pass in a `--export-options-plist` file, which they need to create themselves, or know how to yank off a previously-build IPA.  Flutter's [iOS distribution docs](https://docs.flutter.dev/deployment/ios#create-a-build-archive-with-xcode) don't say what needs to go into the plist, it just points to Apple docs, or tells them to launch Xcode and create the IPA there.

> 1. Run flutter build ipa to produce a build archive.
>
>  Note: When you export your app at the end of Distribute App, Xcode will create a directory containing an IPA of your app and an ExportOptions.plist file. You can create new IPAs with the same options without launching Xcode by running flutter build ipa --export-options-plist=path/to/ExportOptions.plist. See xcodebuild -h for details about the keys in this property list.
>
> 2. Open build/ios/archive/MyApp.xcarchive in Xcode.

I filed https://github.com/flutter/flutter/issues/97179 to track not requiring `--export-options-plist` for the most common `app-store` and `ad-hoc` methods via `flutter build ipa` flags.  Please comment in that issue if you have any feedback!
